cpp/src/io/orc/writer_impl.cu

```diff
@@ -1104,75 +1060,81 @@ encoded_data encode_columns(orc_table_view const& orc_table,
 
   hostdevice_2dvector<gpu::encoder_chunk_streams> chunk_streams(
     num_columns, segmentation.num_rowgroups(), stream);
-  for (size_t col_idx = 0; col_idx < num_columns; col_idx++) {
-    auto const& column = orc_table.column(col_idx);
-    auto col_streams   = chunk_streams[col_idx];
-    for (auto const& stripe : segmentation.stripes) {
-      for (auto rg_idx_it = stripe.cbegin(); rg_idx_it < stripe.cend(); ++rg_idx_it) {
-        auto const rg_idx = *rg_idx_it;
-        auto const& ck    = chunks[col_idx][rg_idx];
-        auto& strm        = col_streams[rg_idx];
+  // per-stripe, per-stream owning buffers
+  std::vector<std::vector<rmm::device_uvector<uint8_t>>> encoded_data(segmentation.num_stripes());
```

Comment:
Not in this PR, but this reminds me that we could extract the owning buffer strong type here (https://github.com/rapidsai/cudf/blob/e37bddb84be7a3a6915511702739f5b7e87386d7/cpp/include/cudf/io/datasource.hpp#L352) and apply it across cuIO code. Thus the above code would be a 2D vector of `owning_buffer`. This also helps unify data type by using `std::byte` consistently.

Also, owning buffer could be enforced via unique ptr.

Reply:
I don't think I get this one. `rmm::device_uvector` already implies an owning buffer. We know that this type deallocates the memory in the destructor.
In the datasource::buffer we have the distinction between owning and non-owning buffers because we use the abstract buffer type, which does not imply ownership, only the access API to the contained data.

Reply:
> rmm::device_uvector already implies an owning buffer

My intention was to use a strong type like:
```cpp
struct owning_buffer{
  ...
  std::unique_ptr<rmm::device_uvector<uint8_t>> data;
};
```
to explicitly represent owning buffers in cuIO.